SAME PROCESS FOR THIS ALSO AS FOR ORDERS DETAILS 

In [2]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt

In [3]:
df = pd.read_csv("./northwind_orders.csv")
df.head(5)

,order_id,customer_id,employee_id,order_date,required_date,shipped_date,ship_via,freight,ship_name,ship_address,ship_city,ship_region,ship_postal_code,ship_country
0,10248,VINET,5,1996-07-04,1996-08-01,1996-07-16,3,32.38,Vins et alcools Chevalier,59 rue de l'Abbaye,Reims,NaN,51100,France
1,10249,TOMSP,6,1996-07-05,1996-08-16,1996-07-10,1,11.61,Toms Spezialitäten,Luisenstr. 48,Münster,NaN,44087,Germany
2,10250,HANAR,4,1996-07-08,1996-08-05,1996-07-12,2,65.83,Hanari Carnes,"Rua do Paço, 67",Rio de Janeiro,RJ,05454-876,Brazil
3,10251,VICTE,3,1996-07-08,1996-08-05,1996-07-15,1,41.34,Victuailles en stock,"2, rue du Commerce",Lyon,NaN,69004,France
4,10252,SUPRD,4,1996-07-09,1996-08-06,1996-07-11,2,51.30,Suprêmes délices,"Boulevard Tirou, 255",Charleroi,NaN,B-6000,Belgium


In [4]:
df.shape

(830, 14)

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 830 entries, 0 to 829
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          830 non-null    int64  
 1   customer_id       830 non-null    str    
 2   employee_id       830 non-null    int64  
 3   order_date        830 non-null    str    
 4   required_date     830 non-null    str    
 5   shipped_date      809 non-null    str    
 6   ship_via          830 non-null    int64  
 7   freight           830 non-null    float64
 8   ship_name         830 non-null    str    
 9   ship_address      830 non-null    str    
 10  ship_city         830 non-null    str    
 11  ship_region       323 non-null    str    
 12  ship_postal_code  811 non-null    str    
 13  ship_country      830 non-null    str    
dtypes: float64(1), int64(3), str(10)
memory usage: 90.9 KB


In [6]:
df.isnull().sum()

order_id              0
customer_id           0
employee_id           0
order_date            0
required_date         0
shipped_date         21
ship_via              0
freight               0
ship_name             0
ship_address          0
ship_city             0
ship_region         507
ship_postal_code     19
ship_country          0
dtype: int64

In [7]:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
825    False
826    False
827    False
828    False
829    False
Length: 830, dtype: bool

In [8]:
df.tail()

,order_id,customer_id,employee_id,order_date,required_date,shipped_date,ship_via,freight,ship_name,ship_address,ship_city,ship_region,ship_postal_code,ship_country
825,11073,PERIC,2,1998-05-05,1998-06-02,NaN,2,24.95,Pericles Comidas clásicas,Calle Dr. Jorge Cash 321,México D.F.,NaN,05033,Mexico
826,11074,SIMOB,7,1998-05-06,1998-06-03,NaN,2,18.44,Simons bistro,Vinbæltet 34,Kobenhavn,NaN,1734,Denmark
827,11075,RICSU,8,1998-05-06,1998-06-03,NaN,2,6.19,Richter Supermarkt,Starenweg 5,Genève,NaN,1204,Switzerland
828,11076,BONAP,4,1998-05-06,1998-06-03,NaN,2,38.28,Bon app',"12, rue des Bouchers",Marseille,NaN,13008,France
829,11077,RATTC,1,1998-05-06,1998-06-03,NaN,2,8.53,Rattlesnake Canyon Grocery,2817 Milton Dr.,Albuquerque,NM,87110,USA


In [10]:
df.nunique()

order_id            830
customer_id          89
employee_id           9
order_date          480
required_date       454
shipped_date        387
ship_via              3
freight             799
ship_name            90
ship_address         89
ship_city            70
ship_region          19
ship_postal_code     84
ship_country         21
dtype: int64

In [12]:
df['ship_country'].unique()

<StringArray>
[     'France',     'Germany',      'Brazil',     'Belgium', 'Switzerland',
   'Venezuela',     'Austria',      'Mexico',         'USA',      'Sweden',
     'Finland',       'Italy',       'Spain',          'UK',     'Ireland',
    'Portugal',      'Canada',     'Denmark',      'Poland',      'Norway',
   'Argentina']
Length: 21, dtype: str

In [16]:
df['order_id'].min()

np.int64(10248)

In [18]:
df['order_id'].max()

np.int64(11077)

In [21]:
df['order_id'].nunique()

830

In [25]:
df.nunique()

order_id            830
customer_id          89
employee_id           9
order_date          480
required_date       454
shipped_date        387
ship_via              3
freight             799
ship_name            90
ship_address         89
ship_city            70
ship_region          19
ship_postal_code     84
ship_country         21
dtype: int64

In [27]:
df['employee_id'].unique()

array([5, 6, 4, 3, 9, 1, 8, 2, 7])

In [40]:
df['ship_via'].value_counts()

ship_via
2    326
3    255
1    249
Name: count, dtype: int64

In [51]:
df['shipped_date'].head()


0    1996-07-16
1    1996-07-10
2    1996-07-12
3    1996-07-15
4    1996-07-11
Name: shipped_date, dtype: str

In [59]:
df['ship_name'].max()

'Wolski Zajazd'

In [60]:
df['ship_name'].min()

"Alfred's Futterkiste"

In [63]:
df['order_date'].min()

'1996-07-04'

In [66]:
df['order_date'].max()

'1998-05-06'

In [67]:
df['order_date'].value_counts()

order_date
1998-02-26    6
1998-03-03    4
1998-03-06    4
1998-03-11    4
1998-03-16    4
             ..
1997-11-26    1
1997-12-01    1
1997-12-04    1
1997-12-09    1
1997-12-12    1
Name: count, Length: 480, dtype: int64

In [72]:
df['order_date'].value_counts().mean()

np.float64(1.7291666666666667)

In [76]:
df.nunique()

order_id            830
customer_id          89
employee_id           9
order_date          480
required_date       454
shipped_date        387
ship_via              3
freight             799
ship_name            90
ship_address         89
ship_city            70
ship_region          19
ship_postal_code     84
ship_country         21
dtype: int64

In [94]:
df.groupby('customer_id')['order_date'].value_counts().sort_values()

customer_id  order_date
BERGS        1997-08-11    1
ALFKI        1997-10-03    1
             1998-01-15    1
             1998-04-09    1
             1997-08-25    1
                          ..
KOENE        1997-02-25    2
LACOR        1998-03-24    2
SAVEA        1997-10-22    2
LINOD        1998-01-19    2
SAVEA        1998-04-17    2
Name: count, Length: 823, dtype: int64

In [95]:
df.isnull().sum(
)

order_id              0
customer_id           0
employee_id           0
order_date            0
required_date         0
shipped_date         21
ship_via              0
freight               0
ship_name             0
ship_address          0
ship_city             0
ship_region         507
ship_postal_code     19
ship_country          0
dtype: int64

In [106]:
df.dropna(inplace=True)

In [107]:
df.shape

(295, 14)

In [114]:
df.isnull().sum()

order_id            0
customer_id         0
employee_id         0
order_date          0
required_date       0
shipped_date        0
ship_via            0
freight             0
ship_name           0
ship_address        0
ship_city           0
ship_region         0
ship_postal_code    0
ship_country        0
dtype: int64

In [103]:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
825    False
826    False
827    False
828    False
829    False
Length: 830, dtype: bool

In [104]:
df.drop_duplicates(inplace=True)

In [111]:
df.duplicated().sum()

np.int64(0)

In [112]:
df[df.duplicated()]

,order_id,customer_id,employee_id,order_date,required_date,shipped_date,ship_via,freight,ship_name,ship_address,ship_city,ship_region,ship_postal_code,ship_country


before droping null values
df.shape = 830 x 14 
after droping 
df.shape = 295 X 14 

In [110]:
df.nunique()

order_id            295
customer_id          31
employee_id           9
order_date          239
required_date       233
shipped_date        216
ship_via              3
freight             293
ship_name            31
ship_address         31
ship_city            25
ship_region          18
ship_postal_code     31
ship_country          5
dtype: int64

In [124]:
df.tail(5)

,order_id,customer_id,employee_id,order_date,required_date,shipped_date,ship_via,freight,ship_name,ship_address,ship_city,ship_region,ship_postal_code,ship_country
801,11049,GOURL,3,1998-04-24,1998-05-22,1998-05-04,1,8.34,Gourmet Lanchonetes,"Av. Brasil, 442",Campinas,SP,04876-786,Brazil
804,11052,HANAR,3,1998-04-27,1998-05-25,1998-05-01,1,67.26,Hanari Carnes,"Rua do Paço, 67",Rio de Janeiro,RJ,05454-876,Brazil
807,11055,HILAA,7,1998-04-28,1998-05-26,1998-05-05,2,120.92,HILARION-Abastos,Carrera 22 con Ave. Carlos Soublette #8-35,San Cristóbal,Táchira,5022,Venezuela
816,11064,SAVEA,1,1998-05-01,1998-05-29,1998-05-04,1,30.09,Save-a-lot Markets,187 Suffolk Ln.,Boise,ID,83720,USA
818,11066,WHITC,7,1998-05-01,1998-05-29,1998-05-04,2,44.72,White Clover Markets,1029 - 12th Ave. S.,Seattle,WA,98124,USA


In [123]:
df.head(5)

,order_id,customer_id,employee_id,order_date,required_date,shipped_date,ship_via,freight,ship_name,ship_address,ship_city,ship_region,ship_postal_code,ship_country
2,10250,HANAR,4,1996-07-08,1996-08-05,1996-07-12,2,65.83,Hanari Carnes,"Rua do Paço, 67",Rio de Janeiro,RJ,05454-876,Brazil
5,10253,HANAR,3,1996-07-10,1996-07-24,1996-07-16,2,58.17,Hanari Carnes,"Rua do Paço, 67",Rio de Janeiro,RJ,05454-876,Brazil
8,10256,WELLI,3,1996-07-15,1996-08-12,1996-07-17,2,13.97,Wellington Importadora,"Rua do Mercado, 12",Resende,SP,08737-363,Brazil
9,10257,HILAA,4,1996-07-16,1996-08-13,1996-07-22,3,81.91,HILARION-Abastos,Carrera 22 con Ave. Carlos Soublette #8-35,San Cristóbal,Táchira,5022,Venezuela
13,10261,QUEDE,4,1996-07-19,1996-08-16,1996-07-30,2,3.05,Que Delícia,"Rua da Panificadora, 12",Rio de Janeiro,RJ,02389-673,Brazil


In [125]:
df.info()

<class 'pandas.DataFrame'>
Index: 295 entries, 2 to 818
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          295 non-null    int64  
 1   customer_id       295 non-null    str    
 2   employee_id       295 non-null    int64  
 3   order_date        295 non-null    str    
 4   required_date     295 non-null    str    
 5   shipped_date      295 non-null    str    
 6   ship_via          295 non-null    int64  
 7   freight           295 non-null    float64
 8   ship_name         295 non-null    str    
 9   ship_address      295 non-null    str    
 10  ship_city         295 non-null    str    
 11  ship_region       295 non-null    str    
 12  ship_postal_code  295 non-null    str    
 13  ship_country      295 non-null    str    
dtypes: float64(1), int64(3), str(10)
memory usage: 34.6 KB


gpt main data head and dtype ka screenshot share kiya hai, isme order_date, required_date, shipped_date ka type str se date karna hai? and ship_postal_code ko str se int  me karna hai. aur han before droping null values df.shape = 830 x 14,  after droping null values df.shape = 295 X 14  kiya main to ab data ye bacha hai.

In [129]:
df['order_date'] = pd.to_datetime(df['order_date'])
df['order_date']

2     1996-07-08
5     1996-07-10
8     1996-07-15
9     1996-07-16
13    1996-07-19
         ...    
801   1998-04-24
804   1998-04-27
807   1998-04-28
816   1998-05-01
818   1998-05-01
Name: order_date, Length: 295, dtype: datetime64[us]

In [130]:
df['required_date'] = pd.to_datetime(df['required_date'])
df['required_date']

2     1996-08-05
5     1996-07-24
8     1996-08-12
9     1996-08-13
13    1996-08-16
         ...    
801   1998-05-22
804   1998-05-25
807   1998-05-26
816   1998-05-29
818   1998-05-29
Name: required_date, Length: 295, dtype: datetime64[us]

In [134]:
df['shipped_date'] = pd.to_datetime(df['shipped_date'])
df['shipped_date'] 

2     1996-07-12
5     1996-07-16
8     1996-07-17
9     1996-07-22
13    1996-07-30
         ...    
801   1998-05-04
804   1998-05-01
807   1998-05-05
816   1998-05-04
818   1998-05-04
Name: shipped_date, Length: 295, dtype: datetime64[us]

In [135]:
df.dtypes

order_id                     int64
customer_id                    str
employee_id                  int64
order_date          datetime64[us]
required_date       datetime64[us]
shipped_date        datetime64[us]
ship_via                     int64
freight                    float64
ship_name                      str
ship_address                   str
ship_city                      str
ship_region                    str
ship_postal_code               str
ship_country                   str
dtype: object

In [136]:
df.head()

,order_id,customer_id,employee_id,order_date,required_date,shipped_date,ship_via,freight,ship_name,ship_address,ship_city,ship_region,ship_postal_code,ship_country
2,10250,HANAR,4,1996-07-08,1996-08-05,1996-07-12,2,65.83,Hanari Carnes,"Rua do Paço, 67",Rio de Janeiro,RJ,05454-876,Brazil
5,10253,HANAR,3,1996-07-10,1996-07-24,1996-07-16,2,58.17,Hanari Carnes,"Rua do Paço, 67",Rio de Janeiro,RJ,05454-876,Brazil
8,10256,WELLI,3,1996-07-15,1996-08-12,1996-07-17,2,13.97,Wellington Importadora,"Rua do Mercado, 12",Resende,SP,08737-363,Brazil
9,10257,HILAA,4,1996-07-16,1996-08-13,1996-07-22,3,81.91,HILARION-Abastos,Carrera 22 con Ave. Carlos Soublette #8-35,San Cristóbal,Táchira,5022,Venezuela
13,10261,QUEDE,4,1996-07-19,1996-08-16,1996-07-30,2,3.05,Que Delícia,"Rua da Panificadora, 12",Rio de Janeiro,RJ,02389-673,Brazil


In [139]:
df.to_csv("northwind_orders_clean_data.csv", index=False)

CREATE TABLE orders (
    order_id         INT PRIMARY KEY,
    customer_id      VARCHAR(10),
    employee_id      INT,
    order_date       DATE,
    required_date    DATE,
    shipped_date      DATE,
    ship_via         INT,
    freight          DECIMAL(10,2),
    ship_name        VARCHAR(100),
    ship_address     VARCHAR(200),
    ship_city        VARCHAR(50),
    ship_region      VARCHAR(50),
    ship_postal_code VARCHAR(20),
    ship_country     VARCHAR(50)
);
